opnode/node/api.go

```diff
@@ -74,3 +90,174 @@ func toBlockNumArg(number rpc.BlockNumber) string {
+		return nil, ethereum.NotFound
+	}
+
+	lastL2Hash := found.Hash
+	lastL2BlockNum := found.Number
+	var batches []*derive.BatchData
+	var totalBatchSizeBytes uint64
+	var hasLargeNextBatch bool
+	// Now continue fetching the next blocks, and build batches, until we either run out of space, or run out of blocks.
+	for i := found.Number + 1; i < found.Number+MaxL2BlocksPerBatchResponse+1; i++ {
+		l2Block, err := n.client.BlockByNumber(ctx, big.NewInt(int64(i)))
+		if err != nil {
+			if errors.Is(err, ethereum.NotFound) { // block number too high
+				break
+			}
+			return nil, fmt.Errorf("failed to retrieve L2 block by number %d: %v", i, err)
+		}
+		batch, err := l2.BlockToBatch(n.config, l2Block)
+		if err != nil {
+			return nil, fmt.Errorf("failed to convert L2 block %d (%s) to batch: %v", i, l2Block.Hash(), err)
+		}
+		if batch == nil { // empty block, nothing to submit as batch
+			lastL2Hash = l2Block.Hash()
+			lastL2BlockNum = l2Block.Number().Uint64()
+			continue
+		}
+
+		// Encode the single as a batch to get a size estimate. This should
+		// slightly overestimate the size of the final batch, since each
+		// serialization will contribute the bundle version byte that is
+		// typically amortized over the entire bundle.
+		//
+		// TODO(conner): use iterative encoder when switching to calldata
+		// compression.
+		var buf bytes.Buffer
+		err = derive.EncodeBatches(n.config, []*derive.BatchData{batch}, &buf)
+		if err != nil {
+			return nil, fmt.Errorf("failed to encode batch for size estimate: %v", err)
+		}
+
+		nextBatchSizeBytes := uint64(len(buf.Bytes()))
+		if totalBatchSizeBytes+nextBatchSizeBytes > uint64(req.MaxSize) {
+			// Adding this batch causes the bundle to be too large. Record
+			// whether the bundle size without the batch fails to meet the
+			// minimum size constraint. This is used below to determine whether
+			// or not to ignore the minimum size check, since in this scnario it
+			// can't be avoided, and the batch submitter must submit the
+			// undersized batch to avoid live locking.
+			hasLargeNextBatch = totalBatchSizeBytes < uint64(req.MinSize)
+			break
+		}
+
+		totalBatchSizeBytes += nextBatchSizeBytes
+		batches = append(batches, batch)
+		lastL2Hash = l2Block.Hash()
+		lastL2BlockNum = l2Block.Number().Uint64()
+	}
+
+	var pruneCount int
+	for {
+
+		var buf bytes.Buffer
+		if err := derive.EncodeBatches(n.config, batches, &buf); err != nil {
+			return nil, fmt.Errorf("failed to encode selected batches as bundle: %v", err)
+		}
+
+		bundleSize := uint64(len(buf.Bytes()))
+
+		// Sanity check the bundle size respects the desired maximum. If we have
+		// exceeded the max size, prune the last block. This is very unlikely to
+		// occur since our initial greedy estimate has a very small, bounded
+		// error tolerance, so simply remove the last block and try again.
+		if bundleSize > uint64(req.MaxSize) {
+			batches = batches[:len(batches)-1]
```

Comment:
`lastL2Hash` and `lastL2BlockNum` would be wrong after pruning this last batch

Reply:
indeed, this one of the remaining issues i was planning to tackle after the replay issue is resolved. locally i have panic in this conditional to make sure it's not the cause